web/src/server_events_dispatch.js

```diff
@@ -689,6 +690,17 @@ export function dispatch_normal_event(event) {
                 // present in the backend/Jinja2 templates.
                 settings_display.set_default_language_name(event.language_name);
             }
+            if (event.property === "default_view") {
+                // If user changed the default_view in the settings, we should set
+                // hash_before_overlay to the other default_view when the overlay is closed.
```

Comment:
Okay.

Reply:
OK -- I think I'll rewrite this to use the original default view rather than so much hardcoding, but I get the idea.

Reply:
It's pretty ugly because of https://github.com/zulip/zulip/issues/25781, but should be a nicer result that doesn't require any changes when we add additional possible default views like inbox.

```
diff --git a/web/src/server_events_dispatch.js b/web/src/server_events_dispatch.js
index a0a37eb0d3..f08fc6cf57 100644
--- a/web/src/server_events_dispatch.js
+++ b/web/src/server_events_dispatch.js
@@ -678,6 +678,7 @@ export function dispatch_normal_event(event) {
                 "send_read_receipts",
             ];
 
+            const original_default_view = user_settings.default_view;
             if (user_display_settings.includes(event.property)) {
                 user_settings[event.property] = event.value;
             }
@@ -690,17 +691,19 @@ export function dispatch_normal_event(event) {
                 // present in the backend/Jinja2 templates.
                 settings_display.set_default_language_name(event.language_name);
             }
-            if (event.property === "default_view") {
-                // If current hash is empty (default view) and user opens settings overlay to change the default
-                // view, then going back to an empty hash on closing the overlay will not be correct since the
-                // default view has changed. So, we try to set hash to the previous value of default view.
-                const new_hash = event.value;
-                if (!browser_history.state.hash_before_overlay && overlays.settings_open()) {
-                    browser_history.state.hash_before_overlay = "#recent";
-                    if (new_hash === "recent_topics") {
-                        browser_history.state.hash_before_overlay = "#all_messages";
-                    }
-                }
+            if (
+                event.property === "default_view" && // If current hash is empty (default view), and the
+                // user changes the default view while in settings,
+                // then going back to an empty hash on closing the
+                // overlay will not match the view currently displayed
+                // under settings, so we set the hash to the previous
+                // value of the default view.
+                !browser_history.state.hash_before_overlay &&
+                overlays.settings_open()
+            ) {
+                browser_history.state.hash_before_overlay =
+                    "#" +
+                    (original_default_view === "recent_topics" ? "recent" : original_default_view);
             }
             if (event.property === "twenty_four_hour_time") {
                 // Rerender the whole message list UI
```